proposals/XXXX-spirv-resource-representation.md

```diff
@@ -0,0 +1,271 @@
+they follow the rules required by the SPIR-V specification. This will be the
+same as the
+[`InvocationInterlockPlacementPass`](https://github.com/KhronosGroup/SPIRV-Tools/blob/682bcd51548e670811f1d03511968bb59a1157ce/source/opt/invocation_interlock_placement_pass.h)
+pass in SPIR-V Tools.
+
+The types for the handles will be target types that represent pointers. The
+handle for the array will be
+
+```llvm-ir
+%T = type { ... } ; Fully laid out version of T.
+%T1 = type { [0 x %T] } ; The SPIR-V backend should turn the array into a runtime array.
+target("spirv.Type", target(spirv.Literal, StorageClass), %T1,
+/* OpTypePointer */32)
+```
+
+The types for the buffers must have an
+[explicit layout](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#ExplicitLayout).
+The layout information will be obtained from the DataLayout class:
+
+1.  Struct offsets will come from `DataLayout::getStructLayout`, which returns
+    the offset for each member.
+2.  The array stride will be the size of the array elements. This assumes that
+    structs have appropriate padding at the end to ensure its size is a multiple
+    of its alignment.
+3.  Matrix stride?
+4.  Row major vs Col major?
+
+It is Clang's responsibility to make sure that the data layout is set correctly,
+and that the structs have the correct explicit padding for this to be correct.
+
+The type of the handle for the counter will be
+
+```llvm-ir
+target("spirv.Type", target(spirv.Literal, /* StorageBuffer */ 12),
+target("spirv.DecoratedType", { i32 }, /* block */ 2),
+/* OpTypePointer */32)
+```
+
+### Textures and typed buffers
+
+All of these resource types are represented using an image type in SPIRV. The
+`Texture*` types are implemented as sampled images. The `RWTexture*` types are
+implemented as storage images. `Buffer` is implemented as a uniform buffer, and
+`RWBuffer` is implemented as a storage buffer.
+
+For these cases the return type from `@llvm.spv.handle.fromBinding` would be the
+image type matching the resource type:
+
+```llvm-ir
+target("spirv.Image", ...)
+```
+
+The details of the `spirv.Image` type depend on the specific declaration. Except
+for the image format, the value for each operand is given in the
+[Mapping Resource Attributes to DXIL and SPIR-V](https://github.com/llvm/wg-hlsl/blob/main/proposals/0015-resource-attributes-in-dxil-and-spirv.md)
+proposal. For all resource types other than `RWBuffer<T>` and `RWTexture*<T>`,
+the image format will be `Unknown`.
+
+For `RWBuffer<T>` and `RWTexture*<T>` resource types, if the Vulkan version is
+1.3 or later, the image format will be `Unknown`. This satisfies
+[VUID-RuntimeSpirv-apiVersion-07954](https://docs.vulkan.org/spec/latest/appendices/spirvenv.html#VUID-RuntimeSpirv-apiVersion-07954)
+and
+[VUID-RuntimeSpirv-apiVersion-07955](https://docs.vulkan.org/spec/latest/appendices/spirvenv.html#VUID-RuntimeSpirv-apiVersion-07954).
```

Comment:
Do you want to allow `vk::image_format` in VK 1.3+ still? Or always emit `Unknown`?

Reply:
That is a good question. We can deal with that when we implement `vk::image_format`. I'd want feedback from IHVs to know if there is any advantage to using a specific format or not.

Reply:
I added an open question for this.